pkg/controller/workload/cache/workload_cache.go

```diff
@@ -28,26 +28,89 @@ import (
 type WorkloadCache interface {
 	GetWorkloadByUid(uid string) *workloadapi.Workload
 	GetWorkloadByAddr(networkAddress NetworkAddress) *workloadapi.Workload
-	AddWorkload(workload *workloadapi.Workload)
+	AddOrUpdateWorkload(workload *workloadapi.Workload) (deletedServices []string, newServices []string)
 	DeleteWorkload(uid string)
 	List() []*workloadapi.Workload
+	GetRelationShip(workloadId uint32, serviceId uint32) (uint32, bool)
+	UpdateRelationShip(workloadId uint32, serviceId uint32, relationId uint32)
+	DeleteRelationShip(serviceId uint32, relationId uint32)
 }
 
 type NetworkAddress struct {
 	Network string
 	Address netip.Addr
 }
 
+type ServiceRelationShipByWorkload struct {
+	workloadId uint32
+	serviceId  uint32
+}
+
+type ServiceRelationShipById struct {
+	serviceId  uint32
+	relationId uint32
+}
+
 type cache struct {
-	byUid  map[string]*workloadapi.Workload
-	byAddr map[NetworkAddress]*workloadapi.Workload
-	mutex  sync.RWMutex
+	byUid                  map[string]*workloadapi.Workload
+	byAddr                 map[NetworkAddress]*workloadapi.Workload
+	relationShipByWorkload map[ServiceRelationShipByWorkload]uint32
+	relationShipById       map[ServiceRelationShipById]uint32
+	mutex                  sync.RWMutex
 }
 
 func NewWorkloadCache() *cache {
 	return &cache{
-		byUid:  make(map[string]*workloadapi.Workload),
-		byAddr: make(map[NetworkAddress]*workloadapi.Workload),
+		byUid:                  make(map[string]*workloadapi.Workload),
+		byAddr:                 make(map[NetworkAddress]*workloadapi.Workload),
+		relationShipByWorkload: make(map[ServiceRelationShipByWorkload]uint32),
+		relationShipById:       make(map[ServiceRelationShipById]uint32),
+	}
+}
+
+func (w *cache) GetRelationShip(workloadId uint32, serviceId uint32) (uint32, bool) {
```

Comment:
what does the return value mean

Reply:
`relationId` corresponds to `backend_index` in the endpoint table.